torch::stable::Tensor beginnings, mainly mem mgmt

janeyx99 opened the pull request:
```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
// the C shim Tensor APIs. We've modeled this class after TensorBase, as custom
// op kernels only really need to interact with Tensor metadata (think sizes,
// strides, device, dtype). Other functions on Tensor (like empty_like) should
// live like the ATen op that they are and exist outside of this struct.
//
// There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the
//    C APIs with AtenTensorHandle. Under the hood we still call to these C shim
//    APIs to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2
//    different functions, e.g., doing something like c = a(t) + b(t) for
//    stable::Tensor t. Thus, we use a shared_ptr here.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works
- includes the results of a battle with template specialization

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #156126
* #155977
* __->__ #155367

janeyx99 committed:
[DRAFT] currently segfaults, i need to figure out the memory management

[ghstack-poisoned]

janeyx99 committed:
Update on "[DRAFT] currently segfaults, i need to figure out the memory management"




[ghstack-poisoned]

swolchok commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,136 @@
+
+  void* data_ptr() const {
+    void* data_ptr;
+    aoti_torch_get_data_ptr(ath_.get(), &data_ptr);
```
don't you need to AOTI_TORCH_CODE_ERROR_CHECK here (and all the other places you call an aoti_torch function)? there is no guarantee there isn't a different implementation of the ABI (such as libtorch from 2030) that may actually throw an error.

janeyx99 replied:
o yes

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

Context can be found in this comment:
```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around the C
// shim Tensor APIs. There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the C
//    APIs with AtenTensorHandle. Under the hood we still call to these C shim APIs
//    to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2 different
//    functions, e.g., doing something like c = a(t) + b(t) if a and b both require
//    ownership of t. Here, we use a shared_ptr accompanied with some other tricks
//    to make our contract-based memory management much more possible.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works




[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,136 @@
+#include <torch/csrc/stable/library.h>
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+
+namespace torch::stable {
```
stable = libtorch ABI = uses the C shim APIs + header only APIs, can rely on libtorch.so during runtime
torch/standalone = header only APIs, no dependency on libtorch.so

ezyang commented:
It's been a while since I've looked at this so I've lost the context. All of this looks plausible. I don't remember what guiding principles you have for whether or not an API gets put into the Tensor API, and how much API coverage you need for this to be "good"?

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

Context can be found in this comment:
```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around the C
// shim Tensor APIs. There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the C
//    APIs with AtenTensorHandle. Under the hood we still call to these C shim APIs
//    to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2 different
//    functions, e.g., doing something like c = a(t) + b(t) if a and b both require
//    ownership of t. Here, we use a shared_ptr accompanied with some other tricks
//    to make our contract-based memory management much more possible.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works




[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,136 @@
+    // The following is our way of incrementing the refcount of the underlying
+    // at::Tensor that we point to. Why do we want this weird behavior?
+    // Because! We expect users to only need a StableIValue when they are trying
+    // to pass the Tensor into a stack-based API, e,g., aoti_torch_call_dispatcher.
```
Offline conclusion: I am going to try to write this function as a part of the `from/to` implementations for StableIValue directly.

desertfire commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,155 @@
+    // has ownership of its Tensor arguments. Since this torch::stable::Tensor
+    // object will likely go out of scope by the end of the user extension's
+    // local function and will thus delete its reference on the at::Tensor,
+    // we create a new AtenTensorHandle for that use case.
```
I think another possible use case of StableIValue is to pass as input parameters to native runtime. The ownership needs to be sorted out for that use case as well.

janeyx99 replied:
Okay, implemented

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

Context can be found in this comment:
```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around the C
// shim Tensor APIs. There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the C
//    APIs with AtenTensorHandle. Under the hood we still call to these C shim APIs
//    to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2 different
//    functions, e.g., doing something like c = a(t) + b(t) if a and b both require
//    ownership of t. Here, we use a shared_ptr accompanied with some other tricks
//    to make our contract-based memory management much more possible.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works




[ghstack-poisoned]

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

Context can be found in this comment:
```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around the C
// shim Tensor APIs. There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the C
//    APIs with AtenTensorHandle. Under the hood we still call to these C shim APIs
//    to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2 different
//    functions, e.g., doing something like c = a(t) + b(t) if a and b both require
//    ownership of t. Here, we use a shared_ptr accompanied with some other tricks
//    to make our contract-based memory management much more possible.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works




[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,113 @@
+  template <typename T>
+  T* data_ptr() const {
+    // the actual implementation directs through mutable_data_ptr but like. how
+    // different is that really.
```
Oh wait! To be clear, this current implementation does call into
```
  void* data_ptr() const {
    return mutable_data_ptr();
  }
  ```
  
  and then static_casts. I think the real difference is that the void* data_ptr ultimately calls into TensorImpl's mutable_data() but the T* data_ptr calls into https://github.com/pytorch/pytorch/blob/2585960/c10/core/TensorImpl.h#L1567-L1576 which performs different checks.

janeyx99 replied:
I ended up removing this api from the PR, it isn’t strictly needed

albanD replied:
We just need to be forward compatible here: we will need const-correct APIs here to properly get the behavior we need from COW. So as long as what we provide doesn't break that we're good.
Indeed, void* calling mutable data_ptr is good.
cc @vfdev-5 I would be curious about your recommendation on what is the minimal set we should expose here?
I expect one mutable and one non-mutable. Recommending people use the non-mutable one if possible and NEVER const-cast them?

vfdev-5 replied:
@albanD maybe cc @VitalyFedyunin ?

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -24,7 +24,10 @@
         def setUpClass(cls):
-            install_cpp_extension(extension_root=Path(__file__).parent.parent)
+            try:
+                import libtorch_agnostic  # noqa: F401
```
Yes but I also want that to be possible, to not rebuild my extension even after torch changed and have it still pass tests! like before!

janeyx99 commented on torch/csrc/stable/library.h:
```diff
@@ -39,13 +66,13 @@ StableIValue from(T val) {
-template <>
-StableIValue from(std::nullopt_t val) {
+// Specialization for std::nullopt_t => StableIValue
+template <typename T>
+std::enable_if_t<std::is_same_v<T, std::nullopt_t>, StableIValue> from(T val) {
```
ok will try! dang I gotta up my LLM prompt skills cuz metamate and Claude did not mention class templating at all. Thanks for the code pointer!

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -222,3 +224,53 @@ STABLE_TORCH_LIBRARY_FRAGMENT(libtorch_agnostic, m) {
+  aoti_torch_call_dispatcher("aten::neg", "", stack1);
+
+  StableIValue stack3[2];
+  stack3[0] = stack1[0];
+  stack3[1] = stack2[0];
```
wait how would u avoid the copy?

swolchok replied:
I'm just being picky. 

```
Tensor divide_neg_exp(Tensor t) {
  StableIValue stack[2];
  stack[0] = from(t);
  stack[1] = from(t);
  aoti_torch_call_dispatcher("aten::exp", "", &stack[1]);
  aoti_torch_call_dispatcher("aten::neg", "", &stack[0]);
  aoti_torch_call_dispatcher("aten::divide", "Tensor", &stack[0]);
  return to<Tensor>(stack[0]);
}
```

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
// the C shim Tensor APIs. We've modeled this class after TensorBase, as custom
// op kernels only really need to interact with Tensor metadata (think sizes,
// strides, device, dtype). Other functions on Tensor (like empty_like) should
// live like the ATen op that they are and exist outside of this struct.
//
// There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the
//    C APIs with AtenTensorHandle. Under the hood we still call to these C shim
//    APIs to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2
//    different functions, e.g., doing something like c = a(t) + b(t) for
//    stable::Tensor t. Thus, we use a shared_ptr here.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works
- includes the results of a battle with template specialization




[ghstack-poisoned]

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
// the C shim Tensor APIs. We've modeled this class after TensorBase, as custom
// op kernels only really need to interact with Tensor metadata (think sizes,
// strides, device, dtype). Other functions on Tensor (like empty_like) should
// live like the ATen op that they are and exist outside of this struct.
//
// There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the
//    C APIs with AtenTensorHandle. Under the hood we still call to these C shim
//    APIs to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2
//    different functions, e.g., doing something like c = a(t) + b(t) for
//    stable::Tensor t. Thus, we use a shared_ptr here.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works
- includes the results of a battle with template specialization




[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/library.h:
```diff
@@ -75,18 +102,40 @@ StableIValue from(std::nullopt_t val) {
+// the input torch::stable::Tensor object will likely go out of scope by
+// the end of the user extension's local function and will thus delete
+// its references on the at::Tensor, we create a new AtenTensorHandle
+// (and thus a new at::Tensor) for that use case.
```
Done here https://github.com/pytorch/pytorch/pull/155984

albanD commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,106 @@
+using DeviceIndex =
+    int8_t; // this is from c10/core/Device.h and can be header only
+
+inline void delete_tensor_object(AtenTensorHandle ath) {
```
Is this intended to be public API?

janeyx99 replied:
added the underscore

janeyx99 replied:
and moved it to namespace detail

albanD commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,106 @@
+  AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_delete_tensor_object(ath));
+}
+
+// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
```
```suggestion
// The torch::stable::Tensor class is a highlevel C++ wrapper around
```
To reduce confusion with standalone

albanD commented on torch/csrc/stable/tensor.h:
```diff
@@ -0,0 +1,106 @@
+  Tensor() = delete;
+
+  // Construct a stable::Tensor from an AtenTensorHandle (ATH)
+  // Steals ownership from the pointer and will be in charge of deleting the
+  // underlying at::Tensor that the ATH points to
```
```suggestion
  // Steals ownership from the ATH
```
I would suggest going for minimal doc for these things to make sure that the key information is very visible at a glance.
The fact that the ATH will be deleted by the destructor is just an implication

albanD reviewed (approved):
Sounds pretty good!

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
// the C shim Tensor APIs. We've modeled this class after TensorBase, as custom
// op kernels only really need to interact with Tensor metadata (think sizes,
// strides, device, dtype). Other functions on Tensor (like empty_like) should
// live like the ATen op that they are and exist outside of this struct.
//
// There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the
//    C APIs with AtenTensorHandle. Under the hood we still call to these C shim
//    APIs to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2
//    different functions, e.g., doing something like c = a(t) + b(t) for
//    stable::Tensor t. Thus, we use a shared_ptr here.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works
- includes the results of a battle with template specialization




[ghstack-poisoned]

janeyx99 committed:
Update on "torch::stable::Tensor beginnings, mainly mem mgmt"

```
// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
// the C shim Tensor APIs. We've modeled this class after TensorBase, as custom
// op kernels only really need to interact with Tensor metadata (think sizes,
// strides, device, dtype). Other functions on Tensor (like empty_like) should
// live like the ATen op that they are and exist outside of this struct.
//
// There are several goals of this class over AtenTensorHandle and
// RAIIAtenTensorHandle:
// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the
//    C APIs with AtenTensorHandle. Under the hood we still call to these C shim
//    APIs to preserve stability.
// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
//    around ownership. This makes it difficult to pass one input into 2
//    different functions, e.g., doing something like c = a(t) + b(t) for
//    stable::Tensor t. Thus, we use a shared_ptr here.
```

This PR:
- exemplifies the above
- adds test cases in libtorch_agnostic to make sure the file actually works
- includes the results of a battle with template specialization




[ghstack-poisoned]
